BUGFIX: Remove users personal workspace on user deletion

mhsdesign opened the pull request:
With https://github.com/neos/neos-development-collection/pull/5146 indeed most problems outlined in https://github.com/neos/neos-development-collection/issues/4566 were solved. Now one non-feature point is kindof left open:

> a user Foo has a personal workspace named user-Foo. If the user is deleted and a new user named Foo is created it will use the previously still existent workspace user-Foo.

In the current state this will no longer happen as `WorkspaceService::getPersonalWorkspaceForUser` operates on the neos UUID user identifier and not the name of the account. Recreating a new user with the same name would mean it will get the workspace `Foo2`. So the only part that is not "fixed" yet is that deleting a user does not cleanup its workspaces.

This change adds a query to determine all content repositories a user was active in and then proceeds to delete the workspaces.

<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
BUGFIX: Remove users personal workspace on user deletion

With https://github.com/neos/neos-development-collection/pull/5146 indeed most problems outlined in https://github.com/neos/neos-development-collection/issues/4566 were solved. Now one non-feature point is kindof left open:

> a user Foo has a personal workspace named user-Foo. If the user is deleted and a new user named Foo is created it will use the previously still existent workspace user-Foo.

In the current state this will no longer happen as `WorkspaceService::getPersonalWorkspaceForUser` operates on the neos UUID user identifier and not the name of the account. Recreating a new user with the same name would mean it will get the workspace `Foo2`. So the only part that is not "fixed" yet is that deleting a user does not cleanup its workspaces.

This change adds a query to determine all content repositories a user was active in and then proceeds to delete the workspaces.

mhsdesign commented:
An alternative - possible more stable solution would be: Let the workspaces be orphaned for a while and have a cleanup command to take care of them. Doing the deletion across all repositories from the UserService and that one request is dangerous as if any of the crs contain an error the whole deletion fails partly.

A cleanup cr command might seem inevitable after all all alone to rebase all users workspaces for once in a while. That could then check if any unknown user ids are in the workspace metadata and go a head to clean them.

So i like this change but a second after making it im doubting it already :D

nezaniel reviewed (approved):
I agree that personal workspace deletion should happen when users are deleted

+1

mhsdesign commented:
@nezaniel and me discussed that we prefer a synchronous behaviour as provided via this pr for now due to the fact that any commands to cleanup things have to be remembered and actually used once in a while. A cronjob is no requirement for neos and for the case a cr cannot be instantiated or is not up to date, deleting a user is not the major concern :)

Thus lets have this 8.3 like approach for now.

mhsdesign commented:
also as Neos 9 betas have been around for a while we also might need a migration strategy for old user workspace where the users where deleted. Deletion of such "orphaned" workspaces is not possible via CLI or workspace module:

`./flow workspace:delete user-foo`
> Did not delete workspace "user-foo" because it is a personal workspace. Personal workspaces cannot be deleted manually.

I would say we can either offer a migration to handle that or simply respect the `--force` option to always allow deletion of user workspaces:

```
./flow workspace:delete user-foo --force
```

which should impose no problem as the user will get a new workspace assigned after login.

mhsdesign committed:
FEATURE: Allow deletion of personal workspace via `--force` CLI

`flow workspace:delete editor-mceditworth`
> Did not delete workspace "editor-mceditworth" because its the personal workspace of user "Editor McEditworth" with 0 unpublished node(s). Use --force to delete it nevertheless.

Also adds "Owner: Editor McEditworth" to `flow workspace:show`

mhsdesign commented:
Okay we now allow deletion of personal workspace via `--force` CLI

`flow workspace:delete editor-mceditworth`
> Did not delete workspace "editor-mceditworth" because its the personal workspace of user "Editor McEditworth" with 0 unpublished node(s). Use --force to delete it nevertheless.
